### ISSUE TITLE
Alternative threads UX: inline

### ISSUE THREAD

**Issue by HarHarLinks** [vector-im/element-meta#2140]
#### What would you like to do?
i'm trying out labs threads, but they often don't feel great in the right side panel.
I imagine there has been a concept of inline UI for threads at some point.

compact, based on the unfolding reply threads feature of element:
![compact threads](https://user-images.githubusercontent.com/2803622/142244625-c8864e21-e172-4ab2-b282-f5f49c0a3ca2.png)

compacter, more reaction-like
![compacter threads](https://user-images.githubusercontent.com/2803622/142244635-cbba8c03-e38b-4bca-96be-710b1ba8d857.png)

(these screenshots are narrow for demo purposes, but of course the messages would be sentence-length instead of single words)

#### Why would you like to do it?
I feel like switching to the panel
- is clunky & disconnected
- not so readable since the panel is ~20% wide, room list is 20% wide, meanwhile the usual timeline is the remaining 60% wide
- on top the whole panel UX is generally really uncomfortable and threads only make it worse

I get that one idea of using the panel is to be able to view both timelines at once. But in the way it is tone, this isn't useful in my eyes:
- I'd rather just reply normally to keep the flow of the main timeline
- switching between threads is tedious

#### How would you like to achieve it?
2 concepts above.
the idea is
- as long as the thread is short, thread replies are somewhere between an (emoji) reaction and a full blown comment. as such the UI design of the thread messages is extremely compact.
- threads stay in line, no disconnect from the normal timeline. multiple can be visible at once, e.g. if someone makes to points in a discussion, comments to each point can happen "in-line"
- long (> 5 replies?) threads can collapse this box and switch to panel mode. possibly keep showing the last couple replies and/or scrollbox similar to how code blocks now work
- preferably you can pop out a thread similar to the image lightbox.

**Comment by joshuaboniface** [comment]
I 100% support this.

This is how Mattermost does threads, versus the current implementation is how Slack does threads. I will refer to them by these names.

Slack-style I find completely unusable. First, it shuffles threads off into a hidden void separate from the main channel, accessible only in a small sliver of the window. Further, since the thread is "tied" to another message in the timeline, it is lost quickly if there are other ongoing discussions. IMO this runs counter to the stated intention in high-traffic threads too: if you try to follow one, and close it to view another, you've lost the previous one in the traffic of the main channel and it's just gone. It's more cumbersome to use and follow.

In contrast, inline-with-sidebar-filtering is how Mattermost does threads. In my opinion it's vastly superior in every usecase. In slow channels with few replies, it quickly groups messages together as replies and makes filtering them for context easy. In high-traffic channels, it prevents what are effectively side "subrooms" that get isolated from the main chat, ensuring that all talk in a channel is actively part of the channel, while still providing logical grouping and context, as well as isolation in the sidebar (which is, in this model, not a primary location but a filtered view) for those wishing to quickly view the thread. And most importantly, since the message are still part of the main channel, it's easy to quickly jump into and out of threads as new replies come in.

What I'd actually prefer is the ability to switch this on a per-user or per-client basis. I'm sure some people love the Slack style, I personally hate it and would rather not have threads at all than this style. Having the option for the Mattermost style would be preferable to me even if it isn't to others, and I know many who agree with this.

tl;dr Do threads how Mattermost does threads, not how Slack does threads. It's better in every way.

**Comment by HarHarLinks** [comment]
> Thus my suggestion stands to be reviewed at a later point.

After 2 years of threads, I think it is time to come back to this.

> my use case differs quite strongly from the intended use case of multiple high traffic threads in high traffic rooms.

I have in all this time used threads every now and then. I'm probably using threads much more than the average user.
I regularly partake in discussions in fairly active rooms, such as https://matrix.to/#/#thisweekinmatrix:matrix.org or https://matrix.to/#/#element-web:matrix.org, though I don't know if those meet Jano's criteria from my previous comment or still differ.
I have noticed that people rarely use threads, and if a couple of people do, there is usually always someone who either missed that there was a thread at all, or people partaking in the thread but in the main timeline. Without more data, I want to chalk that behavior up to 2 main reasons:
1. Plenty of clients that still don't implement threads at all, or just with a fallback UX, such as the Element X clients as of today.
2. The UX around threads being too difficult to use/discover.

Overall, I think these situations I keep experiencing would probably have been handled better with this proposal.
